src/tools/perf/lib/libperf.c

```diff
@@ -906,31 +890,53 @@ static ucs_status_t ucp_perf_test_fill_params(ucx_perf_params_t *params,
     return UCS_OK;
 }
 
+static ucs_status_ptr_t
+ucp_perf_test_destroy_ep(ucp_ep_h ep, ucp_rkey_h rkey, unsigned index)
+{
+    ucs_status_ptr_t *req;
+
+    if (rkey != NULL) {
+        ucp_rkey_destroy(rkey);
+    }
+
+    if (ep != NULL) {
+        req = ucp_ep_close_nb(ep, UCP_EP_CLOSE_MODE_FLUSH);
+
+        if (UCS_PTR_IS_PTR(req)) {
+            return req;
+        } else if (UCS_PTR_STATUS(req) != UCS_OK) {
+            ucs_warn("failed to close ep %p on thread %d: %s\n", ep, index,
+                     ucs_status_string(UCS_PTR_STATUS(req)));
+        }
+    }
+    return NULL;
+}
+
 static void ucp_perf_test_destroy_eps(ucx_perf_context_t* perf)
 {
     unsigned i, thread_count = perf->params.thread_count;
     unsigned num_in_prog     = 0;
     ucs_status_ptr_t **reqs  = ucs_alloca(thread_count * sizeof(*reqs));
     ucs_status_ptr_t *req;
+    ucs_status_ptr_t *self_req;
     ucs_status_t status;
 
     for (i = 0; i < thread_count; ++i) {
-        if (perf->ucp.tctx[i].perf.ucp.rkey != NULL) {
-            ucp_rkey_destroy(perf->ucp.tctx[i].perf.ucp.rkey);
+        req = ucp_perf_test_destroy_ep(perf->ucp.tctx[i].perf.ucp.ep,
+                                       perf->ucp.tctx[i].perf.ucp.rkey, i);
+        if (req != NULL) {
+            reqs[num_in_prog++] = req;
         }
 
-        if (perf->ucp.tctx[i].perf.ucp.ep != NULL) {
-            req = ucp_ep_close_nb(perf->ucp.tctx[i].perf.ucp.ep,
-                                  UCP_EP_CLOSE_MODE_FLUSH);
+        self_req = ucp_perf_test_destroy_ep(
+                perf->ucp.tctx[i].perf.ucp.self_ep,
+                perf->ucp.tctx[i].perf.ucp.self_send_rkey, i);
 
-            if (UCS_PTR_IS_PTR(req)) {
-                reqs[num_in_prog++] = req;
-            } else if (UCS_PTR_STATUS(req) != UCS_OK) {
-                ucs_warn("failed to close ep %p on thread %d: %s\n",
-                         perf->ucp.tctx[i].perf.ucp.ep, i,
-                         ucs_status_string(UCS_PTR_STATUS(req)));
-            }
+        if (self_req != NULL) {
+            ucp_request_release(self_req);
         }
+
+        ucp_rkey_destroy(perf->ucp.tctx[i].perf.ucp.self_recv_rkey);
```

Comment:
1. do it before destroying the ep.
2. could it be NULL?
3. maybe pass the keys that should be destroyed to ucp_perf_test_destroy_ep - as "const ucp_rkey_h *rkeys, unsigned num_rkeys"